[Wu Zengfu] iP

wuzengfu opened the pull request:
## Echobot

> A CLI APP for Task Management.
Echobot frees your mind of having to remember things you need to do. It's,

- test-based
- easly to learn
- ~~Fast~~ _SUPER FAST_ to use

All you need to do is,

1. download it from [here](https://github.com/wuzengfu/ip/releases/tag/v0.1)
2. double-click it
3. adds your tasks
4. let it manages your tasks for you 😊 

And it is **FREE**!

Features:

- [X] Managing Tasks
- [X] Managing Deadlines
- [X] Persistent Storage
- [ ] Reminders (coming soon)

___

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Echobot {
    public static void main(String[] args) {
        new Echobot().run(args);
    }
}

```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

wuzengfu committed:
Rename chatbot, add UI.java to greet users and then exit

wuzengfu committed:
Add echo method

wuzengfu committed:
Add the ability to store task and display all tasks

wuzengfu committed:
Add the ability to mark tasks as done and undone

wuzengfu committed:
Add Event.java

wuzengfu committed:
Add Deadline.java

wuzengfu committed:
Add support for tracking Event, Todo and Deadline

wuzengfu committed:
Add exception classes

wuzengfu committed:
Add error handling for unexpected inputs

wuzengfu committed:
Move task related classes into "task" package

wuzengfu committed:
Add support for deleting tasks from the list

wuzengfu committed:
Add FileManagement.java

wuzengfu committed:
Add support of persistent storage and loading

wuzengfu committed:
Add automated testing of text

wuzengfu committed:
Merge branch 'branch-Level-7'

wuzengfu committed:
Remove unnecessary constructors from task related classes

wuzengfu committed:
Add datetime support for tasks

wuzengfu committed:
Add functionality to list tasks by date

Users are allowed to use "list /on [date]" command to check by date on those uncompleted tasks that they still have.

wuzengfu committed:
Merge branch 'branch-Level-8'

wuzengfu committed:
Extract command parser

wuzengfu committed:
Merge branch 'A-MoreOOP'

wuzengfu committed:
Put all classes in Echobot package

wuzengfu committed:
Allow users to enter wrong command without exiting

wuzengfu committed:
Merge branch 'A-Packages'

wuzengfu committed:
Merge remote-tracking branch 'origin/add-gradle-support'

wuzengfu committed:
Add gradle support

wuzengfu committed:
Merge branch 'A-Gradle'

wuzengfu committed:
Add JUnit tests for CommandParser.java

wuzengfu committed:
Add JUnit tests for some methods in TaskList.java

wuzengfu committed:
Merge branch 'A-JUnit'

wuzengfu committed:
Modify archiveFileName for shadowJar

wuzengfu committed:
Adds JavaDoc comments

wuzengfu committed:
Merge branch 'branch-A-JavaDoc'

wuzengfu committed:
Supports finding tasks by a keyword

wuzengfu committed:
Merge branch 'branch-Level-9'

wuzengfu committed:
Adds GUI

wuzengfu committed:
Modifies gradle main application entry

wuzengfu committed:
Merge branch 'branch-Level-10'

G-Y-Chan commented on src/main/java/echobot/command/Command.java:
```diff
@@ -0,0 +1,37 @@
+public class Command {
+    protected FileManagement fileManagement;
+    protected TaskList taskList;
+    public enum CommandType {
```
Great job on completing the optional task to implement ENUMS in your project! 👍

G-Y-Chan commented on src/main/java/echobot/command/AddCommand.java:
```diff
@@ -0,0 +1,41 @@
+        this.taskToAdd = new ToDo(false, description);
+    }
+
+    public AddCommand(String description, String deadline) throws InvalidDeadlineFormatException, TaskNameEmptyException, DeadlineEmptyException {
+        this.taskToAdd = new Deadline(false, description, deadline);
+    }
+
+    public AddCommand(String description, String from, String to) throws InvalidDeadlineFormatException, TaskNameEmptyException, EventStartEndDateEmptyException {
+        this.taskToAdd = new Event(false, description, from, to);
+    }
```
Perhaps you could consider putting the exceptions in a new line so that your method header will not be too long?

G-Y-Chan commented on src/main/java/echobot/command/AddCommand.java:
```diff
@@ -0,0 +1,41 @@
+    @Override
+    public CommandResponse execute() throws EchoBotException {
+        taskList.addTask(this.taskToAdd);
+        String response = "Got it. I've added this task:\n\t\t\t\t" + this.taskToAdd + "\n\t\t\tNow you have " + this.taskList.size() + " task(s) in the list.";
```
Perhaps you can split this string into multiple lines for better readability?

G-Y-Chan commented on src/main/java/echobot/command/CommandParser.java:
```diff
@@ -0,0 +1,128 @@
+            arguments = commandMatcher.group("arguments").trim();
+        }
+
+        return switch (command) {
```
Great use of a switch block here as it really helps with the clarity of the return value for each of the cases!

G-Y-Chan commented on src/main/java/echobot/command/CommandParser.java:
```diff
@@ -0,0 +1,128 @@
+        }
+        case AddCommand.EVENT_COMMAND: {
+            description = description.substring(0, description.length() - 1).trim();
+            pattern = Pattern.compile(".*?/from (?<from>\\d{2}-\\d{2}-\\d{4} \\d{2}:\\d{2}) /to (?<to>\\d{2}-\\d{2}-\\d{4} \\d{2}:\\d{2})");
```
Perhaps you could split this pattern into multiple lines for better readability?

G-Y-Chan reviewed (commented):
Overall, your code was extremely well written with very clear class structure in use for different classes that extend from a common parent class.
Your code seems to exceed the recommended character limit per line at certain points, but overall a very minor issue.
Great job on completing the various stretch goals for the different levels as well, and well done on you work so far!

wuzengfu commented:
> Overall, your code was extremely well written with very clear class structure in use for different classes that extend from a common parent class. Your code seems to exceed the recommended character limit per line at certain points, but overall a very minor issue. Great job on completing the various stretch goals for the different levels as well, and well done on you work so far!

I have taken note of those issues that you pointed out. Thank you!

jowhee3011 commented on src/main/java/echobot/command/CommandParser.java:
```diff
@@ -0,0 +1,128 @@
+public class CommandParser {
+    public Command parse(String input) throws EchoBotException {
+        input = input.trim();
+        Pattern pattern = Pattern.compile("(?<command>\\S+)(?<arguments>.*)");
```
Good use of pattern matcher here.

jowhee3011 commented on src/main/java/echobot/command/CommandParser.java:
```diff
@@ -0,0 +1,128 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class CommandParser {
```
Java docs can be included for this class as well as its methods.

jowhee3011 commented on src/main/java/echobot/exception/EchoBotException.java:
```diff
@@ -0,0 +1,11 @@
+package echobot.exception;
+
+public class EchoBotException extends Exception {
```
I like that this exception encapsulates all exceptions related to echobot.

jowhee3011 reviewed (commented):
Overall, LGTM. Your code is thoroughly thought out and well structured. However, you could include java docs to improve its readability.

wuzengfu committed:
Updates gradle to include JavaFx dependencies

wuzengfu committed:
Adds assertions for ListByDateCommand and TaskList

There is no any assertion in the program.

These are added to test assumptions of the two methods.

Adds assertions to check if keyword is valid for findTasksByKeyword method to process and to check if date is blank for the execution of ListByDateCommand.

The date should not be empty, otherwise, the program will not find any task. The keyword for finding tasks should not be empty too.

This is different from exceptions, since the validation is already done before these methods can be called.

wuzengfu committed:
Refactors some codes

wuzengfu committed:
Merge pull request #1 from wuzengfu/branch-A-Assertions

Adds assertions

wuzengfu committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/echobot/command/ListByDateCommand.java

wuzengfu committed:
Merge pull request #2 from wuzengfu/branch-A-CodeQuality

Refactors some codes

wuzengfu committed:
Adds dependencies for JUnit tests

wuzengfu committed:
Fixes some styling issues

wuzengfu committed:
Merge branch 'master' into branch-A-Streams

# Conflicts:
#	src/main/java/echobot/task/TaskList.java

wuzengfu committed:
Merge pull request #3 from wuzengfu/branch-A-Streams

Uses stream

wuzengfu committed:
Creates main.yml for CI

wuzengfu committed:
Merge pull request #4 from wuzengfu/branch-A-CI

Creates main.yml for CI

wuzengfu committed:
Supports undo command

Users are not allowed to undo a command.

An undo command gives users flexibility to make mistakes.

A new UndoCommand class and Undoable interface are added. A stack is added to keep track of those undoable commands.

Using stack maintains the order of commands to be undone due to its first-in-first-out property.

wuzengfu committed:
Merge branch 'branch-C-Undo'

wuzengfu committed:
Add JavaDoc command

wuzengfu committed:
Add Ui.png

wuzengfu committed:
Update sample echobot.txt

wuzengfu committed:
Update user guide

wuzengfu committed:
Remove blank lines

wuzengfu committed:
Fix header comments
